### ISSUE TITLE
Audio inaccuracies found by Ace

### ISSUE THREAD

**Issue by bootsector** [MiSTer-devel/MegaDrive_MiSTer#5]
From Discord:

<img width="993" alt="image" src="https://github.com/MiSTer-devel/MegaDrive_MiSTer/assets/125004/3c7c2214-97d3-4e42-be67-484e833549f0">

<img width="1126" alt="image" src="https://github.com/MiSTer-devel/MegaDrive_MiSTer/assets/125004/5e8800fd-41fe-4ea1-9c63-66e235e78280">

https://cdn.discordapp.com/attachments/1096614378968727652/1147379997313613935/DA__ALL_AVG_A-MD1UTVA3-M192_vs_NukedMD_96KHz_0000.png

**Comment by MP2E** [comment]
I can confirm the high-pitched whine in After Burner II. I tried all of the filter settings, but it exists even with no filter.

Perhaps this is a Nuked MD bug?

**Comment by PhantombrainM** [comment]
Reverting Megadrive Test 2 
- made PSG volume louder

might help. @sorgelig

**Comment by sorgelig** [comment]
Need to be more specific how it's loud. Twice louder?

**Comment by sorgelig** [comment]
As far as i know, different models have different volumes

**Comment by Toryalai1** [comment]
I think it is better if someone can provide mdfourier result from the new core vs original hardware. @Ace9921 can probably help out here

**Comment by davedoc1984** [comment]
> As far as i know, different models have different volumes

Yes Model 2 consoles have a blown out distorted volume.

**Comment by MP2E** [comment]
Ace's comparison and comments are specifically about the Model 1 filter, as compared to a Model 1 VA3 Genesis console.

Looks like PSG is about 2 decibels louder than FM currently, so it would need to be about 63% of current volume, if I did my math right.

**Comment by sorgelig** [comment]
Can he compare volume with "No Filter" option?
It's possible that PSG pre-filter may amplify a little.

**Comment by Ace9921** [comment]
Tried with the filter off using both the YM2612 and YM3438, makes no difference.

YM2612:
![DA__ALL_AVG_A-MD1UTVA3-M192_vs_NukedMD_96KHz_no_filter_0000](https://github.com/MiSTer-devel/MegaDrive_MiSTer/assets/46823896/ac6c1917-ad1c-4ce9-b85f-0b9c68ba2a3c)

YM3438:
![DA__ALL_AVG_A-MD1UTVA3-M192_vs_NukedMD_96KHz_YM3438_no__0000](https://github.com/MiSTer-devel/MegaDrive_MiSTer/assets/46823896/3ee6081a-4691-436f-b37b-8ce0cf9bf972)

This is being compared against the VA3 Genesis model 1 again.

**Comment by bootsector** [comment]
According to @nukeykt tests, MiSTer core sounds little close to a Model 1 VA0, maybe sounding a tiny bit louder on MiSTer.

**Comment by sorgelig** [comment]
I will finish with some additional mappers and will back to this issue.
As i understood i need to lower PSG by 2db.

As for whistle: Genesis core detects PCM playback on PSG and do a shortcut with samples. This makes very clean audio but it doesn't work as original HW. In MD core it works like HW. But HW has some filters applied naturally and artificially while FPGA does only digital part. So it needs additional filter to be applied.

I can add filter if someone will tell me which one is needed. Also i've noticed old filters used in Genesis core don't work well but i'm not sure about parameters. I've used comments to generate same filters using MiSTer's IIR filter, but filtering result is different. May be comments in Genesis filter are outdated.

**Comment by MP2E** [comment]
That makes sense, thanks for clarifying

**Comment by sorgelig** [comment]
Try latest changes in PSG

**Comment by MP2E** [comment]
![DA__ALL_AVG_A-MD1UTVA3-M192_vs_untitled_0000](https://github.com/MiSTer-devel/MegaDrive_MiSTer/assets/167708/85ec1050-9c3f-4f81-b3ad-0a2f5f7d2043)
Model 1 Filter, YM2612 with latest changes.

Looks like PSG is too quiet now

**Comment by MP2E** [comment]
Working on this by making changes to the PSG amplification and then testing in MDFourier. I'll post my changes when I'm finished

**Comment by MP2E** [comment]
Made the change in #9 
![DA__ALL_AVG_A-MD1UTVA3-M192_vs_NukedMD_Test_0000](https://github.com/MiSTer-devel/MegaDrive_MiSTer/assets/167708/de788d89-ba3c-4bcf-8f47-1ccdad7441a7)

**Comment by sorgelig** [comment]
Probably PSG pre-filter need to be changed. I've used 8KHz to workaround that whistle in afterburner II.

**Comment by MP2E** [comment]
![DA__ALL_AVG_A-MD1UTVA3-M192_vs_NukedMD_0907_0000](https://github.com/MiSTer-devel/MegaDrive_MiSTer/assets/167708/245e7465-957a-48d8-be43-f5e4931ff8ed)
Latest changes are very close, PSG is still a little bit low.

**Comment by sorgelig** [comment]
You can try to increase the value further till it will match.

**Comment by MP2E** [comment]
[MegaDrive_20230907.zip](https://github.com/MiSTer-devel/MegaDrive_MiSTer/files/12552805/MegaDrive_20230907.zip)
Just submitted a pull request with a PSG volume change, here's an RBF

**Comment by bootsector** [comment]
Volume looks spot on! Thanks @birdybro to teach me how to read this. And thanks @MP2E to looking at it. Now we just need to resolve that dropping yellow line at the end 🤔

**Comment by birdybro** [comment]
That was part of the temporary workaround from sorgelig, an 8khz filter. The whistling is still there but probably most people can't hear it anymore as above 20khz people struggle to hear anything.

**Comment by sorgelig** [comment]
Need to understand the nature of that whistling. Then see how we can fix it.

**Comment by bootsector** [comment]
@sorgelig can you please remove the 8khz filter you added for silencing the whistle so we can see how MDFourier will plot the graph for higher frequencies? That falling yellow line at the end is the bigger problem, I think.

**Comment by birdybro** [comment]
Root cause thanks to advice from @nukeykt and a little bit of my own research for a couple weeks:

PSG sample rate = 223722Hz (In After Burner II)
SMS FM sample rate = 49715Hz
FM synth sample rate = 53267Hz

When these are all shoved together into the genesis_lpf module without any true pre-filtering, the [Gibbs phenomenon](https://en.wikipedia.org/wiki/Gibbs_phenomenon) combined with the fact that decimation will bring down frequencies that are normally not able to be heard by humans down into the human-hearable range, this likely creates the ringing/whistling artifact.

I think the only way to solve this is to pre-filter the PSG by interpolating first and then decimating in stages down to the FM synth sample rate. Then interpolating them together up to the sample rate of the MD master clock. Since the SMS FM synth is also included in this core and could match with the PSG, then it too should be interpolated and decimated to get it to the MD FM sample rate.

This is the same method that jotego had done with his jt12_genmix module and accompanying modules. I found it hard to read what all of the code did because it's a very different code style than my own, so I rewrote what it did to learn about it.

The number of stages and the depth of these stages of each CIC filter step (either interpolation of decimation) will have to be adjusted to account for DC bias and gain, in order to prevent clipping/distortion.

I'm not an electrical engineer at all so the formulas to get the number of filters and stages I'll need were a bit over my head.

WIP draft PR here --> https://github.com/MiSTer-devel/MegaDrive_MiSTer/pull/19

**Comment by SegaSnatcher** [comment]
Birdy, great work!   Thanks for looking into this.

**Comment by SegaSnatcher** [comment]
Speaking of audio, something I just noticed both on the original Genesis core and the new Megadrive core is that there is a noticeable clicking sound when you switch between YM2612 and 3438 sound chips in the menu.  It is easily heard when you just start the core without a game running.  Not sure if this is already known, but thought it would be worth pointing out.

**Comment by sorgelig** [comment]
One click when you switch between chips?

**Comment by sorgelig** [comment]
> PSG sample rate = 223722Hz (In After Burner II)

This is very variable for PCM, and as a PSG it has no sampling rate.

**Comment by birdybro** [comment]
Yes, the PSG sound that isn't PCM is TDM, which doesn't really have a set sample rate since it's multiple asynchronously controlled instruments.

The sample rate for AB2's PCM is coming from the PSG wire input from the core is what I mean when I say PSG, sorry. The PCM from AB2 is taking advantage of the attenuation registers by rapidly switching them on and off, treating the PSG like a 1-bit DAC.

**Comment by sorgelig** [comment]
Yeah, i know PCM is coming from PSG. Other games doing similar trick may have different sampling rate.

**Comment by sorgelig** [comment]
well, it's possible to add an option to have similar PCM trick as it was done in Genesis core.
Not sure what else you mean by modifying ym7101.
